File: scripts/run_authenticated_test.js
const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');
const admin = require('firebase-admin');

// Load test credentials from functions/.env.local
const envPath = path.join(__dirname, '..', 'functions', '.env.local');
let params = {};
if (fs.existsSync(envPath)){
  const content = fs.readFileSync(envPath, 'utf8');
  content.split(/\r?\n/).forEach(line=>{
    line = line.trim(); if(!line || line.startsWith('#')) return;
    const idx = line.indexOf('='); if(idx === -1) return;
    let k = line.slice(0, idx).trim(); let v = line.slice(idx+1).trim();
    v = v.replace(/^\"|\"$/g, '');
    params[k]=v;
  });
}

const CLIENT_CODE = params.ANGELONE_CLIENT_CODE || 'AABL713311';
const PIN = params.ANGELONE_PASSWORD || '1012';
const TOTP = params.ANGELONE_TOTP_TOKEN || 'AGODKRXZZH6FHMYWMSBIK6KDXQ';

// API key extracted from firebase-debug.log (fallback from env FIREBASE_API_KEY)
const API_KEY = process.env.FIREBASE_API_KEY || 'AIzaSyAg6TvwNWGZwEfCP5ZVbszIIBBFHPaEinc';

(async ()=>{
  try{
    // Initialize Firebase Admin using ADC
    if (!admin.apps.length){
      admin.initializeApp();
    }
    const uid = 'test-runner-' + Math.random().toString(36).slice(2,9);
    console.log('Creating custom token for', uid);
    const customToken = await admin.auth().createCustomToken(uid);
    console.log('Custom token created (len):', customToken.length);

    // Exchange custom token for ID token
    const exchangeUrl = `https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken?key=${API_KEY}`;
    const exchangeRes = await fetch(exchangeUrl, {
      method: 'POST', headers: {'Content-Type':'application/json'}, body: JSON.stringify({token: customToken, returnSecureToken: true})
    });
    const exchangeBody = await exchangeRes.text();
    console.log('Exchange status', exchangeRes.status, exchangeBody.slice(0,1000));
    if (exchangeRes.status !== 200){
      throw new Error('Failed to exchange custom token');
    }
    const exchangeJson = JSON.parse(exchangeBody);
    const idToken = exchangeJson.idToken;
    if (!idToken) throw new Error('No idToken in exchange response');
    console.log('Obtained idToken (len):', idToken.length);

    // POST to hosting endpoint
    const endpoint = 'https://tbsignalstream.web.app/api/directAngelLogin';
    console.log('POSTing to', endpoint);
    const postRes = await fetch(endpoint, {
      method: 'POST', headers: { 'Authorization': 'Bearer ' + idToken, 'Content-Type': 'application/json' },
      body: JSON.stringify({ clientCode: CLIENT_CODE, pin: PIN, totp: TOTP }), timeout: 30000
    });
    const postText = await postRes.text();
    console.log('POST status', postRes.status);
    try { console.log('POST JSON:', JSON.parse(postText)); } catch(e){ console.log('POST body text:', postText.slice(0,1000)); }

  }catch(err){
    console.error('Error during authenticated test:', err);
  }
})();

File: scripts/run_authenticated_test.py
import os
import json
import requests
import firebase_admin
from firebase_admin import credentials, auth

# Load test Angel One credentials from functions/.env.local that exists in the repo
ENV_PATH = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'functions', '.env.local')
params = {}
if os.path.exists(ENV_PATH):
    with open(ENV_PATH, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'): continue
            if '=' in line:
                k, v = line.split('=', 1)
                v = v.strip().strip('"')
                params[k.strip()] = v

CLIENT_CODE = params.get('ANGELONE_CLIENT_CODE', 'AABL713311')
PIN = params.get('ANGELONE_PASSWORD', '1012')
TOTP = params.get('ANGELONE_TOTP_TOKEN', 'AGODKRXZZH6FHMYWMSBIK6KDXQ')

# Extracted Firebase Web API key from firebase-debug.log (found in repo)
# If you prefer to set it via env, set FIREBASE_API_KEY
API_KEY = os.environ.get('FIREBASE_API_KEY') or 'AIzaSyAg6TvwNWGZwEfCP5ZVbszIIBBFHPaEinc'

# Initialize Firebase Admin (Application Default Credentials)
if not firebase_admin._apps:
    try:
        cred = credentials.ApplicationDefault()
        firebase_admin.initialize_app(cred)
    except Exception as e:
        print('Failed to initialize firebase_admin:', e)
        raise

# 1) Create custom token
uid = 'test-runner-' + os.urandom(4).hex()
print('Creating custom token for uid:', uid)
custom_token = auth.create_custom_token(uid).decode('utf-8')
print('Custom token created (length):', len(custom_token))

# 2) Exchange custom token for ID token via REST API
url = f'https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken?key={API_KEY}'
body = {
    'token': custom_token,
    'returnSecureToken': True
}
print('Exchanging custom token for ID token...')
res = requests.post(url, json=body, timeout=20)
print('Exchange status:', res.status_code)
print('Exchange response:', res.text)
if res.status_code != 200:
    raise SystemExit('Failed to exchange custom token for ID token')
res_json = res.json()
id_token = res_json.get('idToken')
if not id_token:
    raise SystemExit('No idToken returned from exchange')
print('Obtained idToken (length):', len(id_token))

# 3) POST authenticated request to hosting endpoint
endpoint = 'https://tbsignalstream.web.app/api/directAngelLogin'
headers = {'Authorization': 'Bearer ' + id_token, 'Content-Type': 'application/json'}
payload = {'clientCode': CLIENT_CODE, 'pin': PIN, 'totp': TOTP}
print('POSTing to', endpoint)
resp = requests.post(endpoint, json=payload, headers=headers, timeout=30)
print('Response status:', resp.status_code)
try:
    print('Response JSON:', resp.json())
except Exception:
    print('Response text:', resp.text[:1000])

# Print helpful next steps
print('\nDone. If the request failed, check Cloud Run logs for `directangellogin` service.')
